File: app/src/main/java/com/tommybart/chicagotraintracker/activities/SearchActivity.java
package com.tommybart.chicagotraintracker.activities;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.SearchView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.tommybart.chicagotraintracker.R;
import com.tommybart.chicagotraintracker.adapters.MarginItemDecoration;
import com.tommybart.chicagotraintracker.adapters.SearchAdapter;
import com.tommybart.chicagotraintracker.data.models.Station;

import java.util.ArrayList;
import java.util.Collections;

import static android.content.ContentValues.TAG;

public class SearchActivity extends AppCompatActivity implements View.OnClickListener {

    private RecyclerView searchRecycler;
    private SearchAdapter searchAdapter;
    private ArrayList<Station> searchResults = new ArrayList<>();

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search);

        searchRecycler = findViewById(R.id.search_recycler);
        searchResults.addAll(ArrivalsActivity.stationData.values());
        searchAdapter = new SearchAdapter(searchResults, this);
        searchRecycler.setAdapter(searchAdapter);
        searchRecycler.setLayoutManager(new LinearLayoutManager(this));
        searchRecycler.addItemDecoration(new MarginItemDecoration(24));

        setTitle("");
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_search, menu);

        MenuItem searchMenuItem = menu.findItem(R.id.app_bar_search);
        SearchView searchView = (SearchView) searchMenuItem.getActionView();
        searchView.setQueryHint("Search stations");
        searchView.setMaxWidth(Integer.MAX_VALUE);
        searchView.setIconifiedByDefault(false);
        searchView.requestFocusFromTouch();

        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                Log.d(TAG, "onQueryTextSubmit: " + query);
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                doFilter(newText);
                return false;
            }
        });

        return super.onCreateOptionsMenu(menu);
    }

    private void doFilter(String query) {
        searchResults.clear();
        query = query.toLowerCase();
        for (Station station: ArrivalsActivity.stationData.values()) {
            if (station.getName().toLowerCase().contains(query)) {
                searchResults.add(station);
            }
        }
        Collections.sort(searchResults);
        searchAdapter.notifyDataSetChanged();

    }

    @Override
    public void onClick(View v) {
        int position = searchRecycler.getChildLayoutPosition(v);
        Station selected = searchResults.get(position);
        sendResult(selected);
    }

    private void sendResult(Station selected) {
        Intent data = new Intent(this, ArrivalsActivity.class);
        data.putExtra("STATION", selected);
        setResult(RESULT_OK, data);
        finish();
    }
}

File: app/src/main/java/com/tommybart/chicagotraintracker/ui/activities/search/SearchActivity.kt
package com.tommybart.chicagotraintracker.ui.activities.search

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.View
import android.widget.SearchView
import androidx.appcompat.app.AppCompatDelegate
import androidx.lifecycle.ViewModelProvider
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.LinearLayoutManager
import com.tommybart.chicagotraintracker.R
import com.tommybart.chicagotraintracker.data.models.Station
import com.tommybart.chicagotraintracker.data.provider.USE_DARK_THEME_PREFERENCE
import com.tommybart.chicagotraintracker.internal.extensions.TAG
import com.tommybart.chicagotraintracker.ui.MarginItemDecoration
import com.tommybart.chicagotraintracker.ui.activities.main.MainActivity
import com.tommybart.chicagotraintracker.ui.base.ScopedActivity
import kotlinx.android.synthetic.main.activity_search.*
import kotlinx.coroutines.launch
import org.kodein.di.Kodein
import org.kodein.di.KodeinAware
import org.kodein.di.android.closestKodein
import org.kodein.di.generic.instance

const val SEARCH_ACTIVITY_REQUEST_CODE = 123
const val STATION_RESULT_EXTRA = "STATION"

class SearchActivity : ScopedActivity(), KodeinAware, SearchView.OnQueryTextListener,
    View.OnClickListener {

    override val kodein: Kodein by closestKodein()
    private val viewModelFactory: SearchViewModelFactory by instance<SearchViewModelFactory>()
    private lateinit var viewModel: SearchViewModel

    private lateinit var searchRecyclerAdapter: SearchRecyclerAdapter
    private lateinit var stationList: List<Station>
    private val stationSearchResults = ArrayList<Station>()

    public override fun onCreate(savedInstanceState: Bundle?) {

        val isNightMode = PreferenceManager.getDefaultSharedPreferences(this)
            .getBoolean(USE_DARK_THEME_PREFERENCE, true)
        if (isNightMode) {
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
        } else {
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
        }

        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_search)

        viewModel = ViewModelProvider(this, viewModelFactory)
            .get(SearchViewModel::class.java)
        bindUi()
    }

    private fun bindUi() {
        searchRecyclerAdapter = SearchRecyclerAdapter(this, stationSearchResults)
        searchRecyclerAdapter.setOnClickListener(this)
        search_activity_recycler.adapter = searchRecyclerAdapter
        search_activity_recycler.layoutManager = LinearLayoutManager(this)
        search_activity_recycler.addItemDecoration(MarginItemDecoration(24))
        title = ""
        launch {
            stationList = viewModel.stationData.await()
            stationSearchResults.addAll(stationList)
            search_activity_grp_refresh.visibility = View.GONE
            searchRecyclerAdapter.notifyDataSetChanged()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.activity_search_menu, menu)
        val searchMenuItem = menu.findItem(R.id.app_bar_search)
        val searchView = searchMenuItem.actionView as SearchView
        searchView.queryHint = "Search stations"
        searchView.maxWidth = Int.MAX_VALUE
        searchView.isIconifiedByDefault = false
        searchView.requestFocusFromTouch()
        searchView.setOnQueryTextListener(this)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onQueryTextSubmit(query: String): Boolean {
        Log.d(TAG, "onQueryTextSubmit: $query")
        return false
    }

    override fun onQueryTextChange(newText: String): Boolean {
        if (this::stationList.isInitialized) {
            doFilter(newText)
        } else {
            Log.d(TAG, "Cannot search: station data is not ready")
        }
        return false
    }

    @SuppressLint("DefaultLocale")
    private fun doFilter(query: String) {
        stationSearchResults.clear()
        stationList.forEach { station ->
            if (station.name.toLowerCase().contains(query.toLowerCase())) {
                stationSearchResults.add(station)
            }
        }
        stationSearchResults.sort()
        searchRecyclerAdapter.notifyDataSetChanged()
    }

    override fun onClick(view: View) {
        val position = search_activity_recycler.getChildLayoutPosition(view)
        val selected = stationSearchResults[position]
        setResultAndFinish(selected)
    }

    private fun setResultAndFinish(selected: Station) {
        val data = Intent(this, MainActivity::class.java)
        data.putExtra(STATION_RESULT_EXTRA, selected)
        setResult(Activity.RESULT_OK, data)
        finish()
    }
}